### PR TITLE
feat: add rectangular and gaussian kernels for mutual information loss

### LINKED ISSUE #3473: [Feature]: Implement Rectangular and Truncated Gaussian Kernels for Mutual Information
### 🚀 Feature Description

Implement `rectangular_kernel` and `truncated_gaussian_kernel` functions and register them in the newly created Kernel Enum within the `kornia.losses.mutual_information` module. This completes the kernel dispatch system introduced in PR #3398.

### 📂 Feature Category

Loss Functions

### 💡 Motivation

Following the merge of the Mutual Information Loss (PR #3398), the Kernel Enum currently only supports Kernel.XU.

As discussed in that PR, @edgarriba suggested using an Enum dispatch system, and @completementgaga has already implemented the base Kernel structure in `kornia.losses.mutual_information.` This feature request tracks the implementation of the standard Rectangular and Gaussian kernels to populate that Enum.

### 💭 Proposed Solution

1. Implement Math Functions:

    - `rectangular_kernel(x, window_radius)`: Returns 1.0 if inside the radius, 0.0 otherwise.

    - `truncated_gaussian_kernel(x, sigma)`: Standard Gaussian implementation.

2. Update Enum: Add RECTANGULAR and GAUSSIAN members to the Kernel class.
3. Testing: Add unit tests to verify these kernels work with the existing batch-consistency logic.

### 🔄 Alternatives Considered

We considered allowing users to pass any custom python function as a kernel. This was rejected during the review of PR #3398 to prevent **"uncontrollable bugs"** and ensure JIT/TorchScript compatibility. The Enum approach is the agreed-upon safe alternative.

### 🎯 Use Cases

- Performance Benchmarking: Researchers can compare if the complex Xu kernel actually yields better alignment than a simple Rectangular kernel for their specific dataset.
- Optimization: Users can select the Rectangular kernel for faster, rough alignment phases where high-precision smoothing is less critical.

### 📝 Additional Context

Reference PR: #3398 (specifically the discussion on kernel dispatch).

Current State: The scaffolding (Kernel Enum) is already merged; this issue is strictly about implementing the missing kernel logic.

### 🤝 Contribution Intent

- [x] I plan to submit a PR to implement this feature
- [ ] I'm requesting this feature but not planning to implement it

### PR DESCRIPTION
**Pull request by zer-art**
## 📝 Description

**⚠️ Issue Link Required**: This PR must be linked to an approved and assigned issue. See [Contributing Guide](https://github.com/kornia/kornia/blob/main/CONTRIBUTING.md#pull-request) for details.

Fixes #3473

**Important**:
- Ensure you are assigned to the linked issue before submitting this PR
- This PR should strictly implement what the linked issue describes
- Do not include changes beyond the scope of the linked issue

---

## 🛠️ Changes Made

- [x] Implemented rectangular_kernel: A standard box filter (returns 1.0 within window, 0.0 outside).
- [x] Implemented truncated_gaussian_kernel: A standard Gaussian implementation (using window_radius as sigma).
- [x] Registered new kernels in the Kernel Enum to enable safe dispatch via EntropyBasedLossBase.
- [x] Updated test_batch_consistency in `tests/losses/test_mutual_information.py` to iterate over all available kernel types **(XU, RECTANGULAR, GAUSSIAN)**.

---

## 🧪 How Was This Tested?

- [x] Unit Tests: Updated `tests/losses/test_mutual_information.py`. The test_batch_consistency now verifies that batched execution matches iterative execution for all 3 kernel types.
- [x] Manual Verification: Verified that the implementation aligns with the mathematical definitions discussed in the linked issue.
- [x] Performance/Edge Cases: Optimized the kernel functions to use torch.where with scalars (avoiding torch.tensor overhead inside hot loops). Reduced test loop count from range(10) to range(3) to prevent CI timeouts while maintaining sufficient coverage.

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist

- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] I have performed a self-review of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.
---

## 💭 Additional Context
This feature completes the kernel dispatch system introduced in PR #3398. It allows users to benchmark the trade-off between the smoother Xu kernel and the faster Rectangular kernel or standard Gaussian smoothing. The implementation uses strict Enum typing as requested by @edgarriba to prevent JIT/TorchScript issues.

### CODE REVIEW

File: kornia/losses/mutual_information.py [hunk 1]
```diff
@@ -49,8 +49,48 @@ def xu_kernel(x: torch.Tensor, window_radius: float = 1.0) -> torch.Tensor:
...
+    Returns:
+        torch.Tensor: transformed signal
+    """
+    sigma = window_radius
+    x_abs = torch.abs(x)
+    mask = x_abs <= window_radius
+
+    gaussian_val = torch.exp(-0.5 * (x / sigma) ** 2) / (sigma * (2 * torch.pi) ** 0.5)
+
+    return torch.where(mask, gaussian_val, 0.0)
```
[the remark this answers is not in the document]
Comment (zer-art) [reply kornia/losses/mutual_information.py:86]: The joint histogram is strictly normalized in _joint_histogram_to_entropies (dividing by the sum), which mathematically cancels out any scalar differences in kernel amplitude. I am keeping the standard kernel definitions to maintain parity with the existing xu_kernel implementation, which also relies on this downstream normalization.

File: tests/losses/test_mutual_information.py [hunk 2]
```diff
@@ -139,43 +140,48 @@ def test_batch_consistency(self, device, dtype):
...
-
-                # 3. Compute Iterative Loss (The "Slow but Correct" way)
-                losses = []
-                normalized_losses = []
-                for i in range(unique_batch_dim_1.shape[0]):
-                    loss = mutual_information_loss(unique_batch_dim_1[i], unique_batch_dim_2[i], num_bins=64)
-                    normalized_loss = normalized_mutual_information_loss(
-                        unique_batch_dim_1[i], unique_batch_dim_2[i], num_bins=64
+            for kernel in [Kernel.xu, Kernel.rectangular, Kernel.gaussian]:
+                for _ in range(3):
```
[the remark this answers is not in the document]
Comment (zer-art) [reply tests/losses/test_mutual_information.py]: The reduction to range(3) is a trade-off to prevent CI timeouts.

Since this test primarily verifies deterministic broadcasting logic (checking if batch dimensions are handled correctly), 15 test cases per kernel (5 dimension configurations × 3 iterations) provide sufficient coverage. Increasing it to range(10) would triple the runtime without significantly increasing the likelihood of catching shape-related bugs.

File: kornia/losses/__init__.py [hunk 3]
```diff
@@ -35,6 +35,7 @@
 from .lovasz_softmax import LovaszSoftmaxLoss, lovasz_softmax_loss
 from .ms_ssim import MS_SSIMLoss
 from .mutual_information import (
+    Kernel,
```
Comment (edgarriba) [inline kornia/losses/__init__.py]: i think we should rename this as MIKernel to be less confusing

File: tests/losses/test_mutual_information.py [hunk 4]
```diff
@@ -139,43 +140,48 @@ def test_batch_consistency(self, device, dtype):
...
-                normalized_loss_batch = normalized_mutual_information_loss(img1, img2, num_bins=64)
-
-                # 3. Compute Iterative Loss (The "Slow but Correct" way)
-                losses = []
-                normalized_losses = []
-                for i in range(unique_batch_dim_1.shape[0]):
-                    loss = mutual_information_loss(unique_batch_dim_1[i], unique_batch_dim_2[i], num_bins=64)
-                    normalized_loss = normalized_mutual_information_loss(
-                        unique_batch_dim_1[i], unique_batch_dim_2[i], num_bins=64
+            for kernel in [Kernel.xu, Kernel.rectangular, Kernel.truncated_gaussian]:
```
Comment (edgarriba) [inline tests/losses/test_mutual_information.py]: use parametrize instead

**Review by edgarriba — CHANGES_REQUESTED** [review]

File: kornia/losses/__init__.py (hunk 3 above)
Comment (zer-art) [reply kornia/losses/__init__.py]: Agreed. I have renamed the class to MIKernel to avoid ambiguity and updated the imports accordingly.

File: tests/losses/test_mutual_information.py (hunk 4 above)
Comment (zer-art) [reply tests/losses/test_mutual_information.py]: **Done**. Refactored the test to use `@pytest.mark.parametrize` which cleans up the nested loops and improves test reporting.
